### PR TITLE
Replace Cardigann Anidex indexer for C# implementation with DDOS bypass

### LINKED ISSUE #6834: [anidex] is not returning results
Tracker: anidex
Search: 720p

The website works well, but if I remove the cookies, I see ddos protection. Then the website works fine. Jackett is returning 0 results.

### PR DESCRIPTION
**Pull request by 6cUbi57z**
Resolves #6834

### CODE REVIEW

**Comment by 6cUbi57z** [comment]
Not since submitting. I'll test it again to try reproduce. 

What action were you doing to produce this and were there any relevant log messages with enhanced logging turned on?

**Comment by garfield69** [comment]
I am getting the error simply by adding anidex indexer to my dashboard active list.
My prior log post was from the enhanced log.

to replicate I suggest deleting your active anidex indexer config and adding a fresh one and see if you can duplicate.

**Comment by 6cUbi57z** [comment]
Sorry about that. It was the result of a bad `string.Split()` when there are no cookies. Should be fixed now and I made some additional improvements to error handling and reliability.

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+            { Name = "Order", Value = "desc" };
+            configData.AddDynamic("orderrequestedfromsite", orderSelect);
+        }
+
+        public string SortBy
+        {
+            get
+            {
+                return ((SelectItem)this.configData.GetDynamic("sortrequestedfromsite")).Value;
+            }
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Remove `this` qualifier and convert to read only expression body for each of these:

ex.
`SortBy => ((SelectItem)configData.GetGetDynamic("sortrequestedfromsite")).Value;`

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+                    { "page", "search" },
+                    { "id", string.Join(",", searchCategories) },
+                    { "group", "0" }, // No group
+                    { "q", query.SearchTerm ?? string.Empty },
+                    { "s", this.SortBy },
+                    { "o", this.Order }
+                };
+
+                // Make search request
+                Uri searchUri = new Uri(this.SiteUri, "?" + queryParameters.GetQueryString());
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Convert to interpolated strings and remove `this`:

`$"{SiteUri}?{queryParameters.GetQueryString()}"`

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+        private IEnumerable<ReleaseInfo> ParseResult(string response)
+        {
+            const string ROW_SELECTOR = "div#content table > tbody > tr";
+
+            try
+            {
+                HtmlParser resultParser = new HtmlParser();
+                IHtmlDocument resultDocument = resultParser.ParseDocument(response);
+                IEnumerable<IElement> rows = resultDocument.QuerySelectorAll(ROW_SELECTOR);
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Use `var` declaration:

```C#
var resultDocument = resultParser.ParseDocument(response);
var rows = resultDocument.QuerySelectorAll(ROW_SELECTOR);
```

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+            {
+                // Check if the cookie already exists, if so exit without doing anything
+                if (this.IsCookiePresent("__ddgu") && this.IsCookiePresent("__ddg1"))
+                {
+                    this.logger.Debug("DDOS Guard cookies are already present. Skipping bypass.");
+                    return;
+                }
+
+                // Make a request to DDoS Guard to get the redirect URL
+                List<KeyValuePair<string, string>> ddosPostData = new List<KeyValuePair<string, string>>
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Use `Dictionary<string, string>` instead of `List`

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+                throw new IOException($"Error while configuring DDoS Guard cookie: {ex}");
+            }
+        }
+
+        private bool IsCookiePresent(string name)
+        {
+            string[] rawCookies = this.CookieHeader.Split(';');
+            IDictionary<string, string> cookies = rawCookies
+                .Where(e => e.Contains('='))
+                .ToDictionary((e) => e.Split('=')[0].Trim(), (e) => e.Split('=')[1].Trim());
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Avoid unnecessary conversion here:

```C#
bool cookiePresent = rawCookies.Where(cookie => cookie.Contains('='))
.Any(cookie => cookie.Split('=')[0].Trim().Equals(name, StringComparison.InvariantCultureIgnoreCase);
```

File: src/Jackett.Common/Indexers/Anidex.cs [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+            }
+        }
+
+        private long ParseLongValueFromRow(IElement row, string propertyName, string selector)
+        {
+            try
+            {
+                string text = this.ParseStringValueFromRow(row, propertyName, selector);
+                long value;
+                if (!long.TryParse(text, out value))
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/Anidex.cs]: Use out var syntax:

`if (!long.TryParse(text, out var value))`

**Comment by ngosang** [comment]
I fixed remaining changes in #7142 (because I can't push in this PR)

**Comment by garfield69** [comment]
Jackett 0.12.1765

**Pull request closed without merging**